MRG: make submodule .git directory for versioneer

matthew-brett opened the pull request:
pip will copy the directory it installs from before running the install
/ wheel build.  If the directory is a submodule, then the copy breaks
the connection with the containing project, so git commands to get the
version will fail.

Fix is to re-create the .git directory for submodules.

Version of fix in https://github.com/matthew-brett/multibuild/pull/12
with thanks to Hassan Kibirige for doing the hard work of diagnosis and finding the general solution.

has2k1 commented on common_utils.sh:
```diff
@@ -239,3 +244,20 @@ function install_run {
+    [ -d "$git_loc" ] && return
+    # Need to recreate .git directory for submodule
+    local origin_url=$(cd "$repo_dir" && git config --get remote.origin.url)
+    git clone --recursive --mirror "$repo_dir" "$repo_dir/.git-full"
```
This creates a `bare` repository and you cannot `git checkout` in a bare repository, it has to be in a `worktree` directory.

The solution is to toggle the `bare` config flag,

```sh
git config --local --bool core.bare false
```

However this cannot be done in `fill_submodule` because later on `git fetch origin` will fail with,

```
fatal: Refusing to fetch into current branch refs/heads/master of non-bare repository
```

A possible solution is `git fetch -u origin`, I have not tried it as git warns against using the `-u` option. The other solution is to toggle the bare flag after `git fetch origin`, even after that we have to `git clean` before `git checkout` to prevent `checkout` from complaining about untracked files.

After all this, a working solution is to modify the `clean_code` function

```sh
    (cd $repo_dir \
        && git fetch origin \
        && git config --local --bool core.bare false \
        && git clean -fxd \
        && git checkout $build_commit \
        && git reset --hard \
        && git submodule update --init --recursive)
```

I have some reservations about this too, since it would be neater  if there were no consequences of the git tricks employed in the `fill_submodule` method spilling into the `clean_code` method.

has2k1 commented:
The tests are inadequate, they do not check the integration of `fiill_submodule` with `clean_code`. I do not know if this is easy to fix.

matthew-brett commented:
I switched to doing a full clone (including working tree), and added a test that I can checkout the master branch in a reconstructed directory - does this work for you?

has2k1 commented:
Yes, that does work. You can merge.

matthew-brett commented:
Thanks for your persistence.
